src/acrcssc/azext_acrcssc/helper/_workflow_status.py

```diff
@@ -0,0 +1,600 @@
+                futures.append(future)
+
+            # Wait for all threads to complete
+            concurrent.futures.wait(futures)
+
+    @staticmethod
+    def from_taskrun(cmd,
+                     taskrun_client,
+                     registry,
+                     scan_taskruns,
+                     patch_taskruns,
+                     progress_indicator=None,
+                     workflow_status_filter=None):
+        WorkflowTaskStatus._retrieve_all_tasklogs(cmd, taskrun_client, registry, scan_taskruns, progress_indicator)
+        all_status = {}
+
+        # we only retrieve logs for scan taskruns in the begining, we will need to retrieved the logs for
+        # failed patch taskruns to populate patch_error_reason. It will be faster to retrieve in a batch
+        failed_patch_tasklog_retrieval = []
+
+        for scan in scan_taskruns:
+            if progress_indicator:
+                progress_indicator.update_progress()
+                if hasattr(progress_indicator, 'hook') and \
+                    hasattr(progress_indicator.hook, 'active_progress') and \
+                        hasattr(progress_indicator.hook.active_progress, 'spinner'):
+                    progress_indicator.hook.active_progress.spinner.step(label=progress_indicator.message)
+            if not hasattr(scan, 'task_log_result'):
+                logger.debug("Scan Taskrun: %s has no logs, silent failure", scan.run_id)
+                continue
+
+            image = WorkflowTaskStatus._get_image_from_tasklog(scan.task_log_result)
+
+            if not image:
+                continue
+
+            # need to check if we have the latest scan task
+            task = scan
+            logs = scan.task_log_result
+
+            if image in all_status:
+                task, logs = WorkflowTaskStatus._latest_task(all_status[image].scan_task,
+                                                             all_status[image].scan_logs,
+                                                             scan, scan.task_log_result)
+            else:
+                all_status[image] = WorkflowTaskStatus(image)
+            all_status[image].scan_task = task
+            all_status[image].scan_logs = logs
+            patch_task_id = all_status[image].get_patch_task_from_scan_tasklog()
+            # missing the patch task id means that the scan either failed, or succeeded and patching is not needed.
+            # this is important, because patching status depends on both the patching task status (if it exists)
+            # and the scan task status
+            if patch_task_id is not None:
+                # it is possible for the patch task to be mentioned in the logs, but the API has not returned the
+                # taskrun for it yet, attempt to retrieve it from client
+                try:
+                    patch_task = next((task for task in patch_taskruns if task.run_id == patch_task_id))
```

Comment:
Considering using a default return to avoid the StopIteration e.g.
patch_task = next((task for task in patch_taskruns if task.run_id == patch_task_id), None)
if not patch_task:
    patch_task = WorkflowTaskStatus._get_missing_taskrun(...)